File: backend/src/main/java/com/finance/services/ReportService.java
package com.finance.services;

import com.finance.dto.ReportDtos.AccountBalanceTrendItem;
import com.finance.dto.ReportDtos.CategorySpendResponse;
import com.finance.dto.ReportDtos.FutureBalancePredictionResponse;
import com.finance.dto.ReportDtos.IncomeExpenseTrendItem;
import com.finance.dto.ReportDtos.InsightItem;
import com.finance.dto.ReportDtos.Point;
import com.finance.entities.Budget;
import com.finance.entities.RecurringFrequency;
import com.finance.entities.RecurringTransaction;
import com.finance.entities.TransactionType;
import com.finance.repositories.AccountRepository;
import com.finance.repositories.BudgetRepository;
import com.finance.repositories.RecurringTransactionRepository;
import com.finance.repositories.TransactionRepository;
import com.finance.security.CurrentUserService;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDate;
import java.time.YearMonth;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class ReportService {

    private final TransactionRepository transactionRepository;
    private final AccountRepository accountRepository;
    private final BudgetRepository budgetRepository;
    private final RecurringTransactionRepository recurringTransactionRepository;
    private final CurrentUserService currentUserService;

    @Transactional(readOnly = true)
    public CategorySpendResponse categorySpend(LocalDate startDate, LocalDate endDate) {
        UUID userId = currentUserService.getCurrentUserId();
        var range = normalizeRange(startDate, endDate);
        Map<String, BigDecimal> grouped = new LinkedHashMap<>();
        transactionRepository.findByUserIdAndTransactionDateBetween(userId, range[0], range[1]).stream()
                .filter(tx -> tx.getType() == TransactionType.EXPENSE)
                .forEach(tx -> grouped.merge(
                        tx.getCategory() == null ? "Uncategorized" : tx.getCategory().getName(),
                        tx.getAmount(),
                        BigDecimal::add
                ));
        List<Map<String, Object>> items = grouped.entrySet().stream()
                .sorted(Map.Entry.<String, BigDecimal>comparingByValue().reversed())
                .map(entry -> Map.<String, Object>of("category", entry.getKey(), "amount", entry.getValue()))
                .toList();
        BigDecimal total = grouped.values().stream().reduce(BigDecimal.ZERO, BigDecimal::add);
        return new CategorySpendResponse(items, total);
    }

    @Transactional(readOnly = true)
    public List<IncomeExpenseTrendItem> incomeExpenseTrend(LocalDate startDate, LocalDate endDate) {
        UUID userId = currentUserService.getCurrentUserId();
        var range = normalizeRange(startDate, endDate);
        Map<LocalDate, IncomeExpenseTrendAccumulator> grouped = new LinkedHashMap<>();
        transactionRepository.findByUserIdAndTransactionDateBetween(userId, range[0], range[1]).stream()
                .sorted(Comparator.comparing(com.finance.entities.Transaction::getTransactionDate))
                .forEach(tx -> {
                    if (tx.getType() == TransactionType.INCOME) {
                        IncomeExpenseTrendAccumulator accumulator = grouped.computeIfAbsent(
                                tx.getTransactionDate(),
                                date -> new IncomeExpenseTrendAccumulator()
                        );
                        accumulator.income = accumulator.income.add(tx.getAmount());
                    } else if (tx.getType() == TransactionType.EXPENSE) {
                        IncomeExpenseTrendAccumulator accumulator = grouped.computeIfAbsent(
                                tx.getTransactionDate(),
                                date -> new IncomeExpenseTrendAccumulator()
                        );
                        accumulator.expense = accumulator.expense.add(tx.getAmount());
                    }
                });

        return grouped.entrySet().stream()
                .map(entry -> new IncomeExpenseTrendItem(entry.getKey(), entry.getValue().income, entry.getValue().expense))
                .toList();
    }

    @Transactional(readOnly = true)
    public List<AccountBalanceTrendItem> accountBalanceTrend(LocalDate startDate, LocalDate endDate) {
        UUID userId = currentUserService.getCurrentUserId();
        var range = normalizeRange(startDate, endDate);
        return accountRepository.findByUserIdOrderByCreatedAtDesc(userId).stream()
                .map(account -> {
                    BigDecimal running = account.getOpeningBalance();
                    List<Point> points = new ArrayList<>();
                    var transactions = transactionRepository.findByUserIdAndAccountIdOrderByTransactionDateAsc(userId, account.getId());
                    for (var tx : transactions) {
                        if (tx.getType() == TransactionType.INCOME || tx.getType() == TransactionType.TRANSFER_IN) {
                            running = running.add(tx.getAmount());
                        } else {
                            running = running.subtract(tx.getAmount());
                        }
                        if (!tx.getTransactionDate().isBefore(range[0]) && !tx.getTransactionDate().isAfter(range[1])) {
                            points.add(new Point(tx.getTransactionDate(), running));
                        }
                    }
                    if (points.isEmpty()) {
                        points.add(new Point(range[0], running));
                    }
                    return new AccountBalanceTrendItem(account.getName(), points);
                })
                .toList();
    }

    @Transactional(readOnly = true)
    public List<InsightItem> insights() {
        UUID userId = currentUserService.getCurrentUserId();
        YearMonth currentMonth = YearMonth.now();
        YearMonth previousMonth = currentMonth.minusMonths(1);
        LocalDate currentStart = currentMonth.atDay(1);
        LocalDate currentEnd = currentMonth.atEndOfMonth();
        LocalDate previousStart = previousMonth.atDay(1);
        LocalDate previousEnd = previousMonth.atEndOfMonth();

        var currentTransactions = transactionRepository.findByUserIdAndTransactionDateBetween(userId, currentStart, currentEnd);

        List<InsightItem> insights = new ArrayList<>();
        addOverspendingInsight(insights, currentTransactions);
        addBudgetInsights(insights, userId, currentMonth, currentStart, currentEnd);
        addMonthOverMonthInsight(insights, userId, currentStart, currentEnd, previousStart, previousEnd);

        ensureMinimumInsights(insights);

        return insights.stream().limit(5).toList();
    }

    @Transactional(readOnly = true)
    public FutureBalancePredictionResponse futureBalancePrediction() {
        UUID userId = currentUserService.getCurrentUserId();
        int horizonDays = 30;
        LocalDate today = LocalDate.now();
        LocalDate horizonEnd = today.plusDays(horizonDays - 1L);

        BigDecimal currentBalance = accountRepository.findByUserIdOrderByCreatedAtDesc(userId).stream()
                .map(account -> account.getCurrentBalance() == null ? BigDecimal.ZERO : account.getCurrentBalance())
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        BigDecimal projectedRecurringNet = recurringTransactionRepository.findByUserIdOrderByNextRunDateAsc(userId).stream()
                .map(recurring -> projectedRecurringContribution(recurring, today, horizonEnd))
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        LocalDate spendStart = today.minusDays(horizonDays - 1L);
        BigDecimal totalRecentExpense = transactionRepository.findByUserIdAndTransactionDateBetween(userId, spendStart, today).stream()
                .filter(tx -> tx.getType() == TransactionType.EXPENSE)
                .map(com.finance.entities.Transaction::getAmount)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        BigDecimal averageDailySpending = totalRecentExpense
                .divide(BigDecimal.valueOf(horizonDays), 2, RoundingMode.HALF_UP);

        BigDecimal projectedVariableExpense = averageDailySpending.multiply(BigDecimal.valueOf(horizonDays));
        BigDecimal predictedBalance = currentBalance
                .add(projectedRecurringNet)
                .subtract(projectedVariableExpense)
                .setScale(2, RoundingMode.HALF_UP);

        return new FutureBalancePredictionResponse(
                currentBalance.setScale(2, RoundingMode.HALF_UP),
                projectedRecurringNet.setScale(2, RoundingMode.HALF_UP),
                averageDailySpending,
                predictedBalance,
                horizonDays
        );
    }

    private BigDecimal projectedRecurringContribution(RecurringTransaction recurring, LocalDate rangeStart, LocalDate rangeEnd) {
        if (recurring.getNextRunDate() == null || recurring.getNextRunDate().isAfter(rangeEnd)) {
            return BigDecimal.ZERO;
        }
        LocalDate effectiveEnd = recurring.getEndDate() == null || recurring.getEndDate().isAfter(rangeEnd)
                ? rangeEnd
                : recurring.getEndDate();
        if (effectiveEnd.isBefore(rangeStart)) {
            return BigDecimal.ZERO;
        }

        LocalDate runDate = recurring.getNextRunDate();
        while (runDate.isBefore(rangeStart)) {
            runDate = nextRunDate(runDate, recurring.getFrequency());
            if (runDate.isAfter(effectiveEnd)) {
                return BigDecimal.ZERO;
            }
        }

        int occurrences = 0;
        while (!runDate.isAfter(effectiveEnd)) {
            occurrences++;
            runDate = nextRunDate(runDate, recurring.getFrequency());
        }

        BigDecimal signedAmount = switch (recurring.getType()) {
            case INCOME, TRANSFER_IN -> recurring.getAmount();
            case EXPENSE, TRANSFER_OUT -> recurring.getAmount().negate();
        };

        return signedAmount.multiply(BigDecimal.valueOf(occurrences));
    }

    private LocalDate nextRunDate(LocalDate date, RecurringFrequency frequency) {
        return switch (frequency) {
            case DAILY -> date.plusDays(1);
            case WEEKLY -> date.plusWeeks(1);
            case MONTHLY -> date.plusMonths(1);
            case YEARLY -> date.plusYears(1);
        };
    }

    private void ensureMinimumInsights(List<InsightItem> insights) {
        List<InsightItem> defaults = List.of(
                new InsightItem(
                        "GENERAL",
                        "No insights yet",
                        "Add more transactions this month to generate personalized insights.",
                        "INFO"
                ),
                new InsightItem(
                        "GENERAL",
                        "Track consistently",
                        "Keep logging expenses weekly for more accurate month-over-month insights.",
                        "INFO"
                ),
                new InsightItem(
                        "GENERAL",
                        "Set category budgets",
                        "Budgets make overspending alerts and insights more useful.",
                        "INFO"
                )
        );

        for (InsightItem fallback : defaults) {
            if (insights.size() >= 3) {
                break;
            }
            boolean exists = insights.stream().anyMatch(item -> item.title().equals(fallback.title()));
            if (!exists) {
                insights.add(fallback);
            }
        }
    }

    private void addOverspendingInsight(List<InsightItem> insights, List<com.finance.entities.Transaction> transactions) {
        Map<String, BigDecimal> expenseByCategory = new LinkedHashMap<>();
        for (var tx : transactions) {
            if (tx.getType() != TransactionType.EXPENSE) {
                continue;
            }
            String categoryName = tx.getCategory() == null ? "Uncategorized" : tx.getCategory().getName();
            expenseByCategory.merge(categoryName, tx.getAmount(), BigDecimal::add);
        }

        expenseByCategory.entrySet().stream()
                .max(Map.Entry.comparingByValue())
                .ifPresent(entry -> insights.add(new InsightItem(
                        "SPENDING",
                        "Top spending category",
                        String.format("%s has the highest spend this month at %s.", entry.getKey(), formatCurrency(entry.getValue())),
                        "MEDIUM"
                )));
    }

    private void addBudgetInsights(
            List<InsightItem> insights,
            UUID userId,
            YearMonth currentMonth,
            LocalDate currentStart,
            LocalDate currentEnd
    ) {
        List<Budget> budgets = budgetRepository.findByUserIdAndMonthAndYear(
                userId,
                currentMonth.getMonthValue(),
                currentMonth.getYear()
        );
        boolean hasRisk = false;
        int riskInsightsAdded = 0;

        for (Budget budget : budgets) {
            BigDecimal spent = transactionRepository.sumAmountByUserIdAndCategoryIdAndTypeAndTransactionDateBetween(
                    userId,
                    budget.getCategory().getId(),
                    TransactionType.EXPENSE,
                    currentStart,
                    currentEnd
            );
            if (budget.getAmount().compareTo(BigDecimal.ZERO) <= 0) {
                continue;
            }

            BigDecimal utilization = spent
                    .multiply(BigDecimal.valueOf(100))
                    .divide(budget.getAmount(), 1, RoundingMode.HALF_UP);

            if (utilization.compareTo(BigDecimal.valueOf(100)) >= 0) {
                hasRisk = true;
                if (riskInsightsAdded < 2) {
                    insights.add(new InsightItem(
                            "BUDGET",
                            "Budget exceeded",
                            String.format(
                                    "%s budget exceeded at %s%% (%s spent).",
                                    budget.getCategory().getName(),
                                    utilization.toPlainString(),
                                    formatCurrency(spent)
                            ),
                            "HIGH"
                    ));
                    riskInsightsAdded++;
                }
            } else if (utilization.compareTo(BigDecimal.valueOf(80)) >= 0) {
                hasRisk = true;
                if (riskInsightsAdded < 2) {
                    insights.add(new InsightItem(
                            "BUDGET",
                            "Budget nearing limit",
                            String.format(
                                    "%s budget is at %s%% (%s of %s).",
                                    budget.getCategory().getName(),
                                    utilization.toPlainString(),
                                    formatCurrency(spent),
                                    formatCurrency(budget.getAmount())
                            ),
                            "MEDIUM"
                    ));
                    riskInsightsAdded++;
                }
            }
        }

        if (!budgets.isEmpty() && !hasRisk) {
            insights.add(new InsightItem(
                    "BUDGET",
                    "Budgets are on track",
                    "No category has crossed 80% of its monthly budget yet.",
                    "LOW"
            ));
        }
    }

    private void addMonthOverMonthInsight(
            List<InsightItem> insights,
            UUID userId,
            LocalDate currentStart,
            LocalDate currentEnd,
            LocalDate previousStart,
            LocalDate previousEnd
    ) {
        BigDecimal currentExpense = totalExpenseForRange(userId, currentStart, currentEnd);
        BigDecimal previousExpense = totalExpenseForRange(userId, previousStart, previousEnd);

        if (currentExpense.compareTo(BigDecimal.ZERO) == 0 && previousExpense.compareTo(BigDecimal.ZERO) == 0) {
            return;
        }

        if (previousExpense.compareTo(BigDecimal.ZERO) == 0) {
            insights.add(new InsightItem(
                    "TREND",
                    "Monthly comparison unavailable",
                    "Previous month has no expense data yet.",
                    "INFO"
            ));
            return;
        }

        BigDecimal delta = currentExpense.subtract(previousExpense);
        BigDecimal percentChange = delta
                .multiply(BigDecimal.valueOf(100))
                .divide(previousExpense, 1, RoundingMode.HALF_UP);

        if (delta.compareTo(BigDecimal.ZERO) > 0) {
            insights.add(new InsightItem(
                    "TREND",
                    "Spending increased",
                    String.format("Expenses are up %s%% vs last month.", percentChange.toPlainString()),
                    "MEDIUM"
            ));
        } else if (delta.compareTo(BigDecimal.ZERO) < 0) {
            insights.add(new InsightItem(
                    "TREND",
                    "Spending decreased",
                    String.format("Expenses are down %s%% vs last month.", percentChange.abs().toPlainString()),
                    "LOW"
            ));
        } else {
            insights.add(new InsightItem(
                    "TREND",
                    "Spending is stable",
                    "Expenses are unchanged compared to last month.",
                    "LOW"
            ));
        }
    }

    private BigDecimal totalExpenseForRange(UUID userId, LocalDate startDate, LocalDate endDate) {
        return transactionRepository.findByUserIdAndTransactionDateBetween(userId, startDate, endDate).stream()
                .filter(tx -> tx.getType() == TransactionType.EXPENSE)
                .map(com.finance.entities.Transaction::getAmount)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }

    private String formatCurrency(BigDecimal amount) {
        return "$" + amount.setScale(2, RoundingMode.HALF_UP).toPlainString();
    }

    private LocalDate[] normalizeRange(LocalDate startDate, LocalDate endDate) {
        if (startDate != null && endDate != null) {
            return new LocalDate[]{startDate, endDate};
        }
        YearMonth current = YearMonth.now();
        return new LocalDate[]{current.atDay(1), current.atEndOfMonth()};
    }

    private static class IncomeExpenseTrendAccumulator {
        private BigDecimal income = BigDecimal.ZERO;
        private BigDecimal expense = BigDecimal.ZERO;
    }
}

File: backend-dotnet/FinanceTracker.Api/Services/ReportService.cs
using FinanceTracker.Api.Contracts;
using FinanceTracker.Api.Data;
using FinanceTracker.Api.Entities;
using Microsoft.EntityFrameworkCore;

namespace FinanceTracker.Api.Services;

public sealed class ReportService(
    FinanceTrackerDbContext dbContext,
    CurrentUserService currentUserService,
    AccessControlLayer accessControlLayer)
{
    public async Task<CategorySpendResponse> CategorySpendAsync(DateOnly? startDate, DateOnly? endDate, CancellationToken cancellationToken)
    {
        var range = NormalizeRange(startDate, endDate);
        var accountIds = await GetAccessibleAccountIdsAsync(cancellationToken);

        var transactions = await dbContext.Transactions
            .AsNoTracking()
            .Include(transaction => transaction.Category)
            .Where(transaction => accountIds.Contains(transaction.AccountId)
                                  && transaction.TransactionDate >= range.start
                                  && transaction.TransactionDate <= range.end)
            .ToListAsync(cancellationToken);

        var grouped = new Dictionary<string, decimal>();
        foreach (var transaction in transactions.Where(transaction => transaction.Type == TransactionType.EXPENSE))
        {
            var categoryName = transaction.Category?.Name ?? "Uncategorized";
            grouped[categoryName] = grouped.TryGetValue(categoryName, out var current)
                ? current + transaction.Amount
                : transaction.Amount;
        }

        var items = grouped
            .OrderByDescending(item => item.Value)
            .Select(item => new CategorySpendItem(item.Key, item.Value))
            .ToList();

        return new CategorySpendResponse(items, grouped.Values.Sum());
    }

    public async Task<IReadOnlyList<IncomeExpenseTrendItem>> IncomeExpenseTrendAsync(DateOnly? startDate, DateOnly? endDate, CancellationToken cancellationToken)
    {
        var range = NormalizeRange(startDate, endDate);
        var accountIds = await GetAccessibleAccountIdsAsync(cancellationToken);

        var transactions = await dbContext.Transactions
            .AsNoTracking()
            .Where(transaction => accountIds.Contains(transaction.AccountId)
                                  && transaction.TransactionDate >= range.start
                                  && transaction.TransactionDate <= range.end)
            .OrderBy(transaction => transaction.TransactionDate)
            .ToListAsync(cancellationToken);

        var grouped = new Dictionary<DateOnly, (decimal income, decimal expense)>();
        foreach (var transaction in transactions)
        {
            if (transaction.Type is not (TransactionType.INCOME or TransactionType.EXPENSE))
            {
                continue;
            }

            (decimal income, decimal expense) current = grouped.TryGetValue(transaction.TransactionDate, out var value)
                ? value
                : (income: 0m, expense: 0m);
            grouped[transaction.TransactionDate] = transaction.Type == TransactionType.INCOME
                ? (current.income + transaction.Amount, current.expense)
                : (current.income, current.expense + transaction.Amount);
        }

        return grouped
            .OrderBy(item => item.Key)
            .Select(item => new IncomeExpenseTrendItem(item.Key, item.Value.income, item.Value.expense))
            .ToList();
    }

    public async Task<IReadOnlyList<AccountBalanceTrendItem>> AccountBalanceTrendAsync(DateOnly? startDate, DateOnly? endDate, CancellationToken cancellationToken)
    {
        var range = NormalizeRange(startDate, endDate);
        var accountIds = await GetAccessibleAccountIdsAsync(cancellationToken);

        var accounts = await dbContext.Accounts
            .AsNoTracking()
            .Where(account => accountIds.Contains(account.Id))
            .OrderByDescending(account => account.CreatedAt)
            .ToListAsync(cancellationToken);

        var items = new List<AccountBalanceTrendItem>(accounts.Count);
        foreach (var account in accounts)
        {
            var running = account.OpeningBalance;
            var balanceAtRangeStart = running;
            var points = new List<Point>();
            var transactions = await dbContext.Transactions
                .AsNoTracking()
                .Where(transaction => transaction.AccountId == account.Id
                                      && transaction.TransactionDate <= range.end)
                .OrderBy(transaction => transaction.TransactionDate)
                .ToListAsync(cancellationToken);

            foreach (var transaction in transactions)
            {
                running += SignedAmount(transaction);

                if (transaction.TransactionDate < range.start)
                {
                    balanceAtRangeStart = running;
                    continue;
                }

                if (transaction.TransactionDate <= range.end)
                {
                    points.Add(new Point(transaction.TransactionDate, running));
                }
            }

            if (points.Count == 0)
            {
                points.Add(new Point(range.start, balanceAtRangeStart));
            }

            items.Add(new AccountBalanceTrendItem(account.Name, points));
        }

        return items;
    }

    public async Task<TrendsResponse> TrendsAsync(DateOnly? startDate, DateOnly? endDate, CancellationToken cancellationToken)
    {
        var range = NormalizeRange(startDate, endDate);
        var accountIds = await GetAccessibleAccountIdsAsync(cancellationToken);
        var incomeExpense = await IncomeExpenseTrendAsync(startDate, endDate, cancellationToken);
        var savings = incomeExpense
            .Select(item => new SavingsTrendPoint(
                item.date,
                item.income,
                item.expense,
                Math.Round(item.income - item.expense, 2, MidpointRounding.AwayFromZero)))
            .ToList();

        var transactions = await dbContext.Transactions
            .AsNoTracking()
            .Include(transaction => transaction.Category)
            .Where(transaction => accountIds.Contains(transaction.AccountId)
                                  && transaction.TransactionDate >= range.start
                                  && transaction.TransactionDate <= range.end
                                  && transaction.Type == TransactionType.EXPENSE)
            .ToListAsync(cancellationToken);

        var topCategories = transactions
            .GroupBy(transaction => transaction.Category?.Name ?? "Uncategorized")
            .Select(group => new
            {
                Category = group.Key,
                Total = group.Sum(transaction => transaction.Amount)
            })
            .OrderByDescending(item => item.Total)
            .Take(4)
            .Select(item => item.Category)
            .ToHashSet(StringComparer.OrdinalIgnoreCase);

        var categoryTrends = transactions
            .Where(transaction => topCategories.Contains(transaction.Category?.Name ?? "Uncategorized"))
            .GroupBy(transaction => transaction.Category?.Name ?? "Uncategorized")
            .OrderBy(group => group.Key)
            .Select(group => new CategoryTrendSeries(
                group.Key,
                group.GroupBy(transaction => transaction.TransactionDate)
                    .OrderBy(item => item.Key)
                    .Select(item => new TrendPoint(
                        item.Key,
                        Math.Round(item.Sum(transaction => transaction.Amount), 2, MidpointRounding.AwayFromZero)))
                    .ToList()))
            .ToList();

        return new TrendsResponse(incomeExpense, savings, categoryTrends);
    }

    public async Task<NetWorthResponse> NetWorthAsync(DateOnly? startDate, DateOnly? endDate, CancellationToken cancellationToken)
    {
        var range = NormalizeRange(startDate, endDate);
        var accountIds = await GetAccessibleAccountIdsAsync(cancellationToken);
        var accounts = await dbContext.Accounts
            .AsNoTracking()
            .Where(account => accountIds.Contains(account.Id))
            .ToListAsync(cancellationToken);

        if (accounts.Count == 0)
        {
            return new NetWorthResponse(0m, [new NetWorthPoint(range.start, 0m)]);
        }

        var transactions = await dbContext.Transactions
            .AsNoTracking()
            .Where(transaction => accountIds.Contains(transaction.AccountId)
                                  && transaction.TransactionDate <= range.end)
            .OrderBy(transaction => transaction.TransactionDate)
            .ToListAsync(cancellationToken);

        var running = accounts.Sum(account => account.OpeningBalance);
        var balanceAtRangeStart = running;
        var deltaByDate = new Dictionary<DateOnly, decimal>();

        foreach (var transaction in transactions)
        {
            running += SignedAmount(transaction);
            if (transaction.TransactionDate < range.start)
            {
                balanceAtRangeStart = running;
                continue;
            }

            deltaByDate[transaction.TransactionDate] = deltaByDate.TryGetValue(transaction.TransactionDate, out var current)
                ? current + SignedAmount(transaction)
                : SignedAmount(transaction);
        }

        var points = new List<NetWorthPoint>();
        var projected = balanceAtRangeStart;
        for (var date = range.start; date <= range.end; date = date.AddDays(1))
        {
            if (deltaByDate.TryGetValue(date, out var delta))
            {
                projected += delta;
            }

            points.Add(new NetWorthPoint(date, Math.Round(projected, 2, MidpointRounding.AwayFromZero)));
        }

        var currentNetWorth = Math.Round(accounts.Sum(account => account.CurrentBalance), 2, MidpointRounding.AwayFromZero);
        return new NetWorthResponse(currentNetWorth, points);
    }

    public async Task<IReadOnlyList<InsightItem>> InsightsAsync(CancellationToken cancellationToken)
    {
        var userId = currentUserService.GetCurrentUserId();
        var accountIds = await GetAccessibleAccountIdsAsync(cancellationToken);
        var today = DateOnly.FromDateTime(DateTime.UtcNow);
        var currentStart = new DateOnly(today.Year, today.Month, 1);
        var currentEnd = new DateOnly(today.Year, today.Month, DateTime.DaysInMonth(today.Year, today.Month));
        var previousMonth = currentStart.AddMonths(-1);
        var previousStart = new DateOnly(previousMonth.Year, previousMonth.Month, 1);
        var previousEnd = new DateOnly(previousMonth.Year, previousMonth.Month, DateTime.DaysInMonth(previousMonth.Year, previousMonth.Month));

        var currentTransactions = await dbContext.Transactions
            .AsNoTracking()
            .Include(transaction => transaction.Category)
            .Where(transaction => accountIds.Contains(transaction.AccountId)
                                  && transaction.TransactionDate >= currentStart
                                  && transaction.TransactionDate <= currentEnd)
            .ToListAsync(cancellationToken);

        var insights = new List<InsightItem>();
        AddOverspendingInsight(insights, currentTransactions);
        await AddBudgetInsightsAsync(insights, userId, today, currentStart, currentEnd, cancellationToken);
        await AddMonthOverMonthInsightAsync(insights, accountIds, currentStart, currentEnd, previousStart, previousEnd, cancellationToken);
        EnsureMinimumInsights(insights);
        return insights.Take(5).ToList();
    }

    public async Task<FutureBalancePredictionResponse> FutureBalancePredictionAsync(CancellationToken cancellationToken)
    {
        const int horizonDays = 30;
        var today = DateOnly.FromDateTime(DateTime.UtcNow);
        var horizonEnd = today.AddDays(horizonDays - 1);
        var accountIds = await GetAccessibleAccountIdsAsync(cancellationToken);

        var currentBalance = await dbContext.Accounts
            .AsNoTracking()
            .Where(account => accountIds.Contains(account.Id))
            .Select(account => (decimal?)account.CurrentBalance)
            .SumAsync(cancellationToken) ?? 0m;

        var recurringTransactions = await dbContext.RecurringTransactions
            .AsNoTracking()
            .Where(recurring => accountIds.Contains(recurring.AccountId))
            .OrderBy(recurring => recurring.NextRunDate)
            .ToListAsync(cancellationToken);

        var projectedRecurringNet = recurringTransactions
            .Select(recurring => ProjectedRecurringContribution(recurring, today, horizonEnd))
            .Aggregate(0m, (current, amount) => current + amount);

        var spendStart = today.AddDays(-(horizonDays - 1));
        var totalRecentExpense = await dbContext.Transactions
            .AsNoTracking()
            .Where(transaction => accountIds.Contains(transaction.AccountId)
                                  && transaction.TransactionDate >= spendStart
                                  && transaction.TransactionDate <= today
                                  && transaction.Type == TransactionType.EXPENSE)
            .Select(transaction => (decimal?)transaction.Amount)
            .SumAsync(cancellationToken) ?? 0m;

        var averageDailySpending = Math.Round(totalRecentExpense / horizonDays, 2, MidpointRounding.AwayFromZero);
        var projectedVariableExpense = averageDailySpending * horizonDays;
        var predictedBalance = Math.Round(currentBalance + projectedRecurringNet - projectedVariableExpense, 2, MidpointRounding.AwayFromZero);

        return new FutureBalancePredictionResponse(
            Math.Round(currentBalance, 2, MidpointRounding.AwayFromZero),
            Math.Round(projectedRecurringNet, 2, MidpointRounding.AwayFromZero),
            averageDailySpending,
            predictedBalance,
            horizonDays
        );
    }

    private static decimal ProjectedRecurringContribution(RecurringTransaction recurring, DateOnly rangeStart, DateOnly rangeEnd)
    {
        if (recurring.NextRunDate > rangeEnd)
        {
            return 0m;
        }

        var effectiveEnd = recurring.EndDate is null || recurring.EndDate > rangeEnd
            ? rangeEnd
            : recurring.EndDate.Value;

        if (effectiveEnd < rangeStart)
        {
            return 0m;
        }

        var runDate = recurring.NextRunDate;
        while (runDate < rangeStart)
        {
            runDate = NextRunDate(runDate, recurring.Frequency);
            if (runDate > effectiveEnd)
            {
                return 0m;
            }
        }

        var occurrences = 0;
        while (runDate <= effectiveEnd)
        {
            occurrences++;
            runDate = NextRunDate(runDate, recurring.Frequency);
        }

        var signedAmount = recurring.Type is TransactionType.INCOME or TransactionType.TRANSFER_IN
            ? recurring.Amount
            : -recurring.Amount;

        return signedAmount * occurrences;
    }

    private static DateOnly NextRunDate(DateOnly date, RecurringFrequency frequency)
    {
        return frequency switch
        {
            RecurringFrequency.DAILY => date.AddDays(1),
            RecurringFrequency.WEEKLY => date.AddDays(7),
            RecurringFrequency.MONTHLY => date.AddMonths(1),
            RecurringFrequency.YEARLY => date.AddYears(1),
            _ => date
        };
    }

    private static void EnsureMinimumInsights(List<InsightItem> insights)
    {
        var defaults = new[]
        {
            new InsightItem("GENERAL", "No insights yet", "Add more transactions this month to generate personalized insights.", "INFO"),
            new InsightItem("GENERAL", "Track consistently", "Keep logging expenses weekly for more accurate month-over-month insights.", "INFO"),
            new InsightItem("GENERAL", "Set category budgets", "Budgets make overspending alerts and insights more useful.", "INFO")
        };

        foreach (var fallback in defaults)
        {
            if (insights.Count >= 3)
            {
                break;
            }

            if (insights.All(item => item.title != fallback.title))
            {
                insights.Add(fallback);
            }
        }
    }

    private static void AddOverspendingInsight(List<InsightItem> insights, IEnumerable<Transaction> transactions)
    {
        var expenseByCategory = new Dictionary<string, decimal>();
        foreach (var transaction in transactions)
        {
            if (transaction.Type != TransactionType.EXPENSE)
            {
                continue;
            }

            var categoryName = transaction.Category?.Name ?? "Uncategorized";
            expenseByCategory[categoryName] = expenseByCategory.TryGetValue(categoryName, out var current)
                ? current + transaction.Amount
                : transaction.Amount;
        }

        var topCategory = expenseByCategory.OrderByDescending(item => item.Value).FirstOrDefault();
        if (topCategory.Key is null)
        {
            return;
        }

        insights.Add(new InsightItem(
            "SPENDING",
            "Top spending category",
            $"{topCategory.Key} has the highest spend this month at {FormatCurrency(topCategory.Value)}.",
            "MEDIUM"));
    }

    private async Task AddBudgetInsightsAsync(
        List<InsightItem> insights,
        Guid userId,
        DateOnly today,
        DateOnly currentStart,
        DateOnly currentEnd,
        CancellationToken cancellationToken)
    {
        var budgets = await dbContext.Budgets
            .AsNoTracking()
            .Include(budget => budget.Category)
            .Where(budget => budget.UserId == userId
                             && budget.Month == today.Month
                             && budget.Year == today.Year)
            .ToListAsync(cancellationToken);

        var hasRisk = false;
        var riskInsightsAdded = 0;

        foreach (var budget in budgets)
        {
            var spent = await dbContext.Transactions
                .Where(transaction => transaction.UserId == userId
                                      && transaction.CategoryId == budget.CategoryId
                                      && transaction.Type == TransactionType.EXPENSE
                                      && transaction.TransactionDate >= currentStart
                                      && transaction.TransactionDate <= currentEnd)
                .Select(transaction => (decimal?)transaction.Amount)
                .SumAsync(cancellationToken) ?? 0m;

            if (budget.Amount <= 0)
            {
                continue;
            }

            var utilization = Math.Round(spent * 100m / budget.Amount, 1, MidpointRounding.AwayFromZero);
            if (utilization >= 100m)
            {
                hasRisk = true;
                if (riskInsightsAdded < 2)
                {
                    insights.Add(new InsightItem(
                        "BUDGET",
                        "Budget exceeded",
                        $"{budget.Category.Name} budget exceeded at {utilization:0.0}% ({FormatCurrency(spent)} spent).",
                        "HIGH"));
                    riskInsightsAdded++;
                }
            }
            else if (utilization >= 80m)
            {
                hasRisk = true;
                if (riskInsightsAdded < 2)
                {
                    insights.Add(new InsightItem(
                        "BUDGET",
                        "Budget nearing limit",
                        $"{budget.Category.Name} budget is at {utilization:0.0}% ({FormatCurrency(spent)} of {FormatCurrency(budget.Amount)}).",
                        "MEDIUM"));
                    riskInsightsAdded++;
                }
            }
        }

        if (budgets.Count != 0 && !hasRisk)
        {
            insights.Add(new InsightItem(
                "BUDGET",
                "Budgets are on track",
                "No category has crossed 80% of its monthly budget yet.",
                "LOW"));
        }
    }

    private async Task AddMonthOverMonthInsightAsync(
        List<InsightItem> insights,
        IReadOnlyList<Guid> accountIds,
        DateOnly currentStart,
        DateOnly currentEnd,
        DateOnly previousStart,
        DateOnly previousEnd,
        CancellationToken cancellationToken)
    {
        var currentExpense = await TotalExpenseForRangeAsync(accountIds, currentStart, currentEnd, cancellationToken);
        var previousExpense = await TotalExpenseForRangeAsync(accountIds, previousStart, previousEnd, cancellationToken);

        if (currentExpense == 0m && previousExpense == 0m)
        {
            return;
        }

        if (previousExpense == 0m)
        {
            insights.Add(new InsightItem(
                "TREND",
                "Monthly comparison unavailable",
                "Previous month has no expense data yet.",
                "INFO"));
            return;
        }

        var delta = currentExpense - previousExpense;
        var percentChange = Math.Round(delta * 100m / previousExpense, 1, MidpointRounding.AwayFromZero);

        if (delta > 0m)
        {
            insights.Add(new InsightItem(
                "TREND",
                "Spending increased",
                $"Expenses are up {percentChange:0.0}% vs last month.",
                "MEDIUM"));
        }
        else if (delta < 0m)
        {
            insights.Add(new InsightItem(
                "TREND",
                "Spending decreased",
                $"Expenses are down {Math.Abs(percentChange):0.0}% vs last month.",
                "LOW"));
        }
        else
        {
            insights.Add(new InsightItem(
                "TREND",
                "Spending is stable",
                "Expenses are unchanged compared to last month.",
                "LOW"));
        }
    }

    private async Task<decimal> TotalExpenseForRangeAsync(IReadOnlyList<Guid> accountIds, DateOnly startDate, DateOnly endDate, CancellationToken cancellationToken)
    {
        return await dbContext.Transactions
            .AsNoTracking()
            .Where(transaction => accountIds.Contains(transaction.AccountId)
                                  && transaction.TransactionDate >= startDate
                                  && transaction.TransactionDate <= endDate
                                  && transaction.Type == TransactionType.EXPENSE)
            .Select(transaction => (decimal?)transaction.Amount)
            .SumAsync(cancellationToken) ?? 0m;
    }

    private async Task<IReadOnlyList<Guid>> GetAccessibleAccountIdsAsync(CancellationToken cancellationToken)
    {
        return await accessControlLayer.GetAccessibleAccountIdsAsync(currentUserService.GetCurrentUserId(), cancellationToken);
    }

    private static decimal SignedAmount(Transaction transaction)
    {
        return transaction.Type is TransactionType.INCOME or TransactionType.TRANSFER_IN
            ? transaction.Amount
            : -transaction.Amount;
    }

    private static string FormatCurrency(decimal amount) => $"${Math.Round(amount, 2, MidpointRounding.AwayFromZero):0.00}";

    private static (DateOnly start, DateOnly end) NormalizeRange(DateOnly? startDate, DateOnly? endDate)
    {
        var today = DateOnly.FromDateTime(DateTime.UtcNow);
        var defaultStart = new DateOnly(today.Year, today.Month, 1);
        var defaultEnd = new DateOnly(today.Year, today.Month, DateTime.DaysInMonth(today.Year, today.Month));
        var start = startDate ?? defaultStart;
        var end = endDate ?? defaultEnd;
        if (start > end)
        {
            (start, end) = (end, start);
        }

        return (start, end);
    }
}
